c/cert/src/rules/INT34-C/ExprShiftedbyNegativeOrGreaterPrecisionOperand.ql

```diff
@@ -0,0 +1,83 @@
+/**
+ * @id c/cert/expr-shiftedby-negative-or-greater-precision-operand
+ * @name INT34-C: Bit shift should not be done by a negative operand or an operand of greater-or-equal precision than that of another
+ * @description Shifting an expression by an operand that is negative or of precision greater or
+ *              equal to that or the another causes representational error.
+ * @kind problem
+ * @precision very-high
+ * @problem.severity error
+ * @tags external/cert/id/int34-c
+ *       external/cert/obligation/rule
+ */
+
+import cpp
+import codingstandards.c.cert
+
+/*
+ * Precision predicate based on a sample implementaion from
+ * https://wiki.sei.cmu.edu/confluence/display/c/INT35-C.+Use+correct+integer+precisions
+ */
+
+int getPrecision(BuiltInType type) {
+  type.(CharType).isExplicitlyUnsigned() and result = 8
+  or
+  type.(ShortType).isExplicitlyUnsigned() and result = 16
+  or
+  type.(IntType).isExplicitlyUnsigned() and result = 32
+  or
+  type.(LongType).isExplicitlyUnsigned() and result = 32
+  or
+  type.(LongLongType).isExplicitlyUnsigned() and result = 64
+  or
+  type instanceof CharType and not type.(CharType).isExplicitlyUnsigned() and result = 7
+  or
+  type instanceof ShortType and not type.(ShortType).isExplicitlyUnsigned() and result = 15
+  or
+  type instanceof IntType and not type.(IntType).isExplicitlyUnsigned() and result = 31
+  or
+  type instanceof LongType and not type.(LongType).isExplicitlyUnsigned() and result = 31
+  or
+  type instanceof LongLongType and not type.(LongLongType).isExplicitlyUnsigned() and result = 63
+}
+
+/* The -1 number literal. */
+class MinusNumberLiteral extends UnaryMinusExpr {
+  MinusNumberLiteral() { this.getOperand() instanceof Literal }
+
+  override string toString() { result = "-" + this.getOperand().toString() }
+}
+
+class ForbiddenShiftExpr extends BinaryBitwiseOperation {
+  ForbiddenShiftExpr() {
+    (
+      /* First Case: Precision mismatch between operands */
+      getPrecision(this.(LShiftExpr).getLeftOperand().getUnderlyingType()) <=
+        getPrecision(this.(LShiftExpr).getRightOperand().getUnderlyingType()) or
```

Comment:
I think we need to compare the precision of the type on the left hand side with the _range of the value_  on the right hand side, e.g.
```
getPrecision(this.(LShiftExpr).getLeftOperand().getUnderlyingType()) < upperBound(this.(LShiftExpr).getRightOperand())
```
And for the negative case:
```
lowerBound(this.(LShiftExpr).getRightOperand()) < 0
```

Reply:
You will also need to consider how to use `GuardCondition`s to identify the "valid" precision guards in the CERT rule text.

Reply:
Addressed with additional check on a negative integer guard check.